core/src/main/java/io/aiven/klaw/service/CacheService.java

```diff
@@ -0,0 +1,158 @@
+package io.aiven.klaw.service;
+
+import java.util.*;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.*;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+@Slf4j
+public class CacheService<T> {
+
+  public static final String URL_SEPERATOR = "/";
+  public static final String TENANT = "tenant";
+  public static final String ID = "id";
+  Map<Integer, Map<Integer, T>> cache;
+
+  RestTemplate rest;
+
+  private String urlEndpoint;
+
+  private HighAvailabilityUtilsService utilsService;
+
+  private String clusterUrlsAsString;
+
+  private List<String> clusterUrls;
+
+  public CacheService(
+      String endpoint, String clusterUrlsAsString, HighAvailabilityUtilsService utilsService) {
+    // TODO an interface needs to be added to allow the passing in of different communication
+    // methods (kafka/https/rabbitmq etc) to allow any org to use what they want for maintaining
+    // cache.
+    this.cache = new HashMap<>();
+    this.urlEndpoint = "cache" + URL_SEPERATOR + endpoint;
+    this.clusterUrlsAsString = clusterUrlsAsString;
+    this.utilsService = utilsService;
+  }
+
+  public T add(int tenantId, Integer id, T entry) {
+    return addOrUpdate(tenantId, id, entry, false);
+  }
+
+  public void addAll(int tenantId, Map<Integer, T> entries) {
+    if (entries == null) {
+      entries = new HashMap<>();
+    }
+    getCache(tenantId).putAll(entries);
+  }
+
+  public T remove(int tenantId, Integer id, boolean isLocalUpdate) {
+    log.info("remove id {}", id);
+    if (!isLocalUpdate) {
+      sendHighAvailabilityRemove(tenantId, id);
+    }
+    return getCache(tenantId).remove(id);
+  }
+
+  public Map<Integer, T> removeCache(int tenantId) {
+    return cache.remove(tenantId);
+  }
+
+  public T update(int tenantId, Integer id, T entry) {
+    return addOrUpdate(tenantId, id, entry, false);
+  }
+
+  public T addOrUpdate(int tenantId, Integer id, T entry, boolean isLocalUpdate) {
+    log.info("addOrUpdate {}", entry);
+    getCache(tenantId).put(id, entry);
+    if (!isLocalUpdate) {
+      sendHighAvailabilityUpdate(tenantId, id, entry);
+    }
+    return entry;
+  }
+
+  public Optional<T> get(int tenantId, Integer id) {
+    return Optional.ofNullable(getCache(tenantId).get(id));
+  }
+
+  public Map<Integer, T> getCache(Integer tenantId) {
```

Comment:
this method seems to be invoked several times than required I think. Can you check this again just in case.